massifs/mmriver.go

```diff
@@ -0,0 +1,246 @@
+	TreeSize2        uint64   `cbor:"2,keyasint"`
+	ConsistencyPaths [][]byte `cbor:"3,keyasint"`
+	RightPeaks       [][]byte `cbor:"4,keyasint"`
+}
+
+type MMRiverVerifiableProofs struct {
+	InclusionProofs   []MMRiverInclusionProof   `cbor:"-1,keyasint,omitempty"`
+	ConsistencyProofs []MMRiverConsistencyProof `cbor:"-2,keyasint,omitempty"`
+}
+
+// MMRiverInclusionProofHeader provides for encoding, and defered decoding, of
+// COSE_Sign1 message headers for MMRIVER receipts
+type MMRiverVerifiableProofsHeader struct {
+	VerifiableProofs MMRiverVerifiableProofs `cbor:"396,keyasint"`
+}
+
+/*
+func SetMMRiverInclusionProofsHeader(
+	msg *commoncose.CoseSign1Message, massif MassifReader, mmrSize, mmrIndex uint64) error {
+	msg.Headers.Unprotected[VDSCoseReceiptProofsTag] = proofs
+}*/
+
+// VerifySignedInclusionReceipts verifies a signed COSE receipt encoded according to the MMRIVER VDS
+// on success the produced root is returned.
+// Signature verification failure is not an error, but the returned root will be nil and the result will be false.
+// All other unexpected issues are returned as errors, with a false result and nil root.
+// Note that MMRIVER receipts allow for multiple inclusion proofs to be attached to the receipt.
+// This function returns true only if ALL receipts verify
+func VerifySignedInclusionReceipts(
+	ctx context.Context,
+	receipt *commoncose.CoseSign1Message,
+	candidates [][]byte,
+) (bool, []byte, error) {
+
+	var err error
+
+	// ignore any existing payload
+	receipt.Payload = nil
+
+	// We must return false if there are no candidates
+	if len(candidates) == 0 {
+		return false, nil, fmt.Errorf("no candidates provided")
+	}
+
+	var header MMRiverVerifiableProofsHeader
+	err = cbor.Unmarshal(receipt.Headers.RawUnprotected, &header)
+	if err != nil {
+		return false, nil, fmt.Errorf("MMRIVER receipt proofs malformed")
+	}
+	verifiableProofs := header.VerifiableProofs
+	if len(verifiableProofs.InclusionProofs) == 0 {
+		return false, nil, fmt.Errorf("MMRIVER receipt inclusion proofs not present")
+	}
+
+	// permit *fewer* candidates than proofs, but not more
+	if len(candidates) > len(verifiableProofs.InclusionProofs) {
+		return false, nil, fmt.Errorf("MMRIVER receipt more candidates than proofs")
+	}
+
+	var proof MMRiverInclusionProof
+
+	proof = verifiableProofs.InclusionProofs[0]
+	receipt.Payload = mmr.IncludedRoot(
+		sha256.New(),
+		proof.Index, candidates[0],
+		proof.InclusionPath)
+
+	err = receipt.VerifyWithCWTPublicKey(nil)
+	if err != nil {
+		return false, nil, fmt.Errorf(
+			"MMRIVER receipt VERIFY FAILED for: mmrIndex %d, candidate %d, err %v", proof.Index, 0, err)
+	}
+	// verify the first proof then just compare the produced roots
+
+	for i := 1; i < len(verifiableProofs.InclusionProofs); i++ {
+
+		proof = verifiableProofs.InclusionProofs[i]
+		proven := mmr.IncludedRoot(sha256.New(), proof.Index, candidates[i], proof.InclusionPath)
+		if bytes.Compare(receipt.Payload, proven) != 0 {
```

Comment:
This means that there is a constraint on all the inclusion paths to always lead to the same accumulator/root?

Which shouldn't be an issue, but i see further down we currently only create a receipt with 1 candidate.

Is it something we wish to do in the future to create receipts with multiple candidates?

Reply:
Yes, a single receipt can contain proofs for many leaves. But only if they share the same accumulator peak. The longer the leaf is in the log, the more siblings it will get. That came for free and has little divergence with more traditional CT style receipts.

We _could_ do a full multi receipt where the payload is a catenation of leaves, but I think that we should wait for a bit before considering that. If I put that in the MMRIVER draft it will almost certainly have some negative feedback.